src/reducers/index.ts

```diff
@@ -52,21 +52,51 @@ function checkStoreInvariants(state: StoreState) {
     state.currentMatchIndex >= -1 &&
       state.currentMatchIndex <= state.matchesList.length
   );
+
+  checkCondition(
+    'every matchId in matchIdToMatchState is also present in matchesList',
+    Object.keys(state.matchIdToMatchState).reduce(
+      (a, e) =>
+        a && state.matchesList.filter(v => v.matchId === e).length === 1,
+      true
+    )
+  );
 }
 
 function reduce(state: StoreState, action: Action) {
   if (action.setGamesList) {
     return { ...state, gamesList: action.setGamesList };
   } else if (action.setMatchesList) {
-    return { ...state, matchesList: action.setMatchesList };
+    let { matchesList, matchIdToMatchState, ...rest } = state;
+    let newMatchIdToMatchState = {};
+    action.setMatchesList.forEach(e => {
+      if (e.matchId in matchIdToMatchState) {
+        newMatchIdToMatchState[e.matchId] = matchIdToMatchState[e.matchId];
+      } else {
+        newMatchIdToMatchState[e.matchId] = {};
+      }
+    });
+    return {
+      matchesList: action.setMatchesList,
+      matchIdToMatchState: newMatchIdToMatchState,
+      ...rest
+    };
   } else if (action.setSignals) {
     return { ...state, signals: action.setSignals };
   } else if (action.setMyUser) {
     return { ...state, myUser: action.setMyUser };
     // TODO: support all other reducers.
   } else if (action.setCurrentMatchIndex) {
-    return {...state, currentMatchIndex: action.setCurrentMatchIndex};
-
+    return { ...state, currentMatchIndex: action.setCurrentMatchIndex };
+  } else if (action.updateMatchIdToMatchState) {
+    let { matchIdToMatchState, ...rest } = state;
```

Comment:
Remove this line and directly pass `...state` instead of rest and use `state[matchIdToMatchState` in second line.

I'm being picky here, both are ok though.